### PR TITLE
[PM-12358] New Verified Organization Domain SSO Detail endpoint

### PR DESCRIPTION
**Pull request by jrmccannon**
## 🎟️ Tracking
[PM-12358](https://bitwarden.atlassian.net/browse/PM-12358)

## 📔 Objective
This is adding a new endpoint behind a feature flag to return only verified domains and their organizations for SSO matching the domain provided via login email. Organizations must have SSO enabled and the domain must be claimed. This has been built to allow for multiple organizations to claim the same domain.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-12358]: https://bitwarden.atlassian.net/browse/PM-12358?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Api/AdminConsole/Controllers/OrganizationDomainController.cs [hunk 1]
```diff
@@ -133,6 +135,20 @@ public async Task<OrganizationDomainSsoDetailsResponseModel> GetOrgDomainSsoDeta
         return new OrganizationDomainSsoDetailsResponseModel(ssoResult);
     }
 
+    [AllowAnonymous]
```
Comment (addisonbeck) [inline src/Api/AdminConsole/Controllers/OrganizationDomainController.cs:138]: Can you explain to me why these need `AllowAnonymous`? I can see it on the method above too, but I don't immediately understand why. Claiming domains is done from authorized sessions in the admin console, isn't it?

File: src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs [hunk 2]
```diff
@@ -95,6 +95,29 @@ from s in sJoin.DefaultIfEmpty()
...
+        using var scope = ServiceScopeFactory.CreateScope();
+        var dbContext = GetDatabaseContext(scope);
+        return await (from o in dbContext.Organizations
+                      from od in o.Domains
+                      join s in dbContext.SsoConfigs on o.Id equals s.OrganizationId into sJoin
+                      from s in sJoin.DefaultIfEmpty()
+                      where od.DomainName == domainName
+                            && o.Enabled
+                            && s.Enabled
+                            && od.VerifiedDate != null
```
Comment (addisonbeck) [inline src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs:111]: This is missing some of the conditions from its MSSQL equivalent, like directly checking that the SSO policy is enabled for the organization.

File: src/Api/AdminConsole/Controllers/OrganizationDomainController.cs (hunk 1 above)
Comment (jrmccannon) [reply src/Api/AdminConsole/Controllers/OrganizationDomainController.cs:138]: This is a get for retrieving the claimed domains. It is used when attempting to log in via SSO. The user is not logged in at this point. This endpoint is retrieved in order to try to determine the org and domain needed for SSO login.

File: src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs (hunk 2 above)
Comment (jrmccannon) [reply src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs:111]: Good point. I looked over what I had in the sproc compared to what was being used, and I see that it doesn't actually care about the policy of `SSO Required`. It just cares if SSO is available (if they have a sso config). So, I adjusted the sproc to reflect that, and that matches the EF query.

**Review by addisonbeck — DISMISSED** [review]
The failing lint job is unrelated and is broken on main. Everything else here looks right to me.

**Review by JaredSnider-Bitwarden — DISMISSED** [review]
LGTM! Nice work!

**Review by rkac-bw — APPROVED** [review]
lgtm

File: util/Migrator/DbScripts/2024-09-26_AddVerifiedOrganizationDomainSsoDetails_ReadByEmail.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    O.Id AS OrganizationId,
+    O.Identifier AS OrganizationIdentifier,
+    OD.DomainName
+FROM [dbo].[OrganizationView] O
+    INNER JOIN [dbo].[OrganizationDomainView] OD ON O.Id = OD.OrganizationId
+    LEFT JOIN [dbo].[Ssoconfig] S ON O.Id = S.OrganizationId
+WHERE OD.DomainName = @Domain
+    AND O.Enabled = 1
+    AND OD.VerifiedDate IS NOT NULL
+    AND (P.Id is NULL OR (P.Id IS NOT NULL AND P.[Type] = 4)) -- SSO Type
```
[the remark this answers is not in the document]
Comment (jrmccannon) [reply util/Migrator/DbScripts/2024-09-26_AddVerifiedOrganizationDomainSsoDetails_ReadByEmail.sql]: That has been removed.

File: src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs [hunk 4]
```diff
@@ -95,6 +95,29 @@ from s in sJoin.DefaultIfEmpty()
...
+    public async Task<IEnumerable<VerifiedOrganizationDomainSsoDetail>> GetVerifiedOrganizationDomainSsoDetailsAsync(string email)
+    {
+        var domainName = new MailAddress(email).Host;
+
+        using var scope = ServiceScopeFactory.CreateScope();
+        var dbContext = GetDatabaseContext(scope);
+        return await (from o in dbContext.Organizations
+                      from od in o.Domains
+                      join s in dbContext.SsoConfigs on o.Id equals s.OrganizationId into sJoin
+                      from s in sJoin.DefaultIfEmpty()
```
[the remark this answers is not in the document]
Comment (jrmccannon) [reply src/Infrastructure.EntityFramework/Repositories/OrganizationDomainRepository.cs:107]: Policy is unneeded for this. The original join to Policy was about checking if SSO was required...which isn't a restraint we need for this query.

**Review by rkac-bw — APPROVED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Review by addisonbeck — APPROVED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Review by addisonbeck — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]
Approving on behalf of @addisonbeck.

**Pull request merged**